### PR TITLE
Allow addition and compound assigment of string and xml

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
This PR fixes https://github.com/ballerina-platform/ballerina-lang/issues/13948 and https://github.com/ballerina-platform/ballerina-lang/issues/14612

```ballerina
xml a = xml `<a>hello</a>`;
string b = "stringLit";
xml c = a + b; // c = <a>hello</a>stringLit
c += "-str-2" // c = <a>hello</a>stringLit-str-2

```

This also fixes xml.length() function returning incorrect number of codePoints when applied to char sequence.

```ballerina
xml a = xml `abc`;
xml b = xml `def`;
xml c = a + b;

int nA = a.length(); // prev: 1 | after this pr: 3
int nB = b.length(); // prev: 1 | after this pr: 3
int nC = c.length(); // prev: 2 | after this pr: 6
```

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/XMLUtils.java [hunk 1]
```diff
@@ -335,7 +335,11 @@ public static BXML tableToXML(BTable table) {
      */
     public static BXML<?> createXMLText(String content) {
         // Remove carriage return on windows environments to eliminate additional &#xd; being added
-        content = content.replace("\r\n", "\n");
+        content = content
+                .replace("\r\n", "\n")
+                .replace("&gt;", ">")
```
Comment (rdhananjaya) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/XMLUtils.java]: <, >, & symbols even though represented as escaped characters in xml (as &lt; &gt; &amp), to xml infoset they are just the plain character (<, >, &). And they are counted  as one character each.

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/statements/compoundassignment/CompoundAssignmentTest.java [hunk 2]
```diff
@@ -321,6 +322,16 @@ public void testCompoundAssignmentAdditionStructElements() {
         Assert.assertEquals(((BInteger) returns[0]).intValue(), 500);
     }
 
+    @Test(description = "Test compound assignment with addition.")
+    public void testCompoundAssignmentOfXmlAndString() {
+        BValue[] returns = BRunUtil.invoke(result, "xmlCompoundAssignment");
+        Assert.assertTrue(returns[0] instanceof BXMLSequence);
+        BXMLSequence sequence = (BXMLSequence) returns[0];
+        Assert.assertEquals(sequence.size(), 4);
```
Comment (MaryamZi) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/statements/compoundassignment/CompoundAssignmentTest.java]: Just to clarify based on this example, according to the spec, "if one operand is a string and one is xml, the string is treated as if it were an xml sequence with one character item for each code point in the string".

So when we are adding "hah" and "blah", is it not like adding two sequences of lengths 3 and 4 respectively?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -2502,6 +2512,16 @@ public void visit(BLangBinaryExpr binaryExpr) {
         }
     }
 
+    private BLangXMLTextLiteral createXMLTextLiteralNode(
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: I think we could leave it here as there is not much of a re-use for this method. I'm OK going with any.

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/add-operation.bal [hunk 4]
```diff
@@ -20,4 +20,20 @@ function intFloatAdd(int a, float b) returns (float) {
 
 function floatIntAdd(float a, int b) returns (float) {
     return a + b;
-}
\ No newline at end of file
+}
+
+function xmlXmlAdd(xml a, xml b) returns (xml) {
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply tests/ballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/add-operation.bal]: I was just following the files convention here.

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/statements/compoundassignment/CompoundAssignmentTest.java (hunk 2 above)
Comment (rdhananjaya) [reply tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/statements/compoundassignment/CompoundAssignmentTest.java]: Yes you are correct, there is a bug in BXMLSequence.size() method. Will fix that.

**Comment by hasithaa** [comment]
@rdhananjaya I have merged, Maryams's external change PR, which caused this to conflict. Can you update you PR?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hasithaa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Seems like a static method, Let's move this to ASTBuilderUtils.

File: composer/package-lock.json [hunk 5]
```diff
@@ -955,8 +955,8 @@
       "integrity": "sha512-r5160ogAvGyHsal38Kux7YYtodEKOj89RGb28ht1jh3SJb08VwRwAKKJL0bGb04Zd/3r9FL3BFIc3bBidYffCA==",
       "dev": true,
       "requires": {
-        "exec-sh": "^0.3.2",
-        "minimist": "^1.2.0"
+        "exec-sh": "0.3.2",
```
Comment (hasithaa) [inline composer/package-lock.json]: Seems like this is added by mistake.  Let's revert this change.

File: tests/ballerina-integration-test/src/test/java/org/ballerinalang/test/socket/SocketBaseTest.java [hunk 6]
```diff
@@ -55,6 +54,7 @@ public void start() throws BallerinaTestException {
         String[] args = new String[] { "-e", "certificate.key=" + privateKey, "-e", "public.cert=" + publicCert };
         serverInstance = new BServerInstance(balServer);
         serverInstance.startServer(balFile, "services", args, requiredPorts);
+        executor.execute(mockSocketServer);
```
Comment (hasithaa) [inline tests/ballerina-integration-test/src/test/java/org/ballerinalang/test/socket/SocketBaseTest.java]: Any reason to do this change? seems like this is not related to XML, string concatenation.

File: tool-plugins/vscode/package-lock.json [hunk 7]
```diff
@@ -257,7 +257,7 @@
     },
     "async": {
       "version": "1.2.1",
-      "resolved": "https://registry.npmjs.org/async/-/async-1.2.1.tgz",
+      "resolved": "http://registry.npmjs.org/async/-/async-1.2.1.tgz",
```
Comment (hasithaa) [inline tool-plugins/vscode/package-lock.json]: Mistakenly committed.

Comment (rdhananjaya) [reply tool-plugins/vscode/package-lock.json]: I think these files are committed with merge. Will re-merge and push
